components/event-sources/reconciler/objects/service.go

```diff
@@ -54,66 +54,84 @@ type ServiceOption func(*servingv1alpha1.Service)
 // WithServiceControllerRef sets the controller reference of a Service.
 func WithServiceControllerRef(or *metav1.OwnerReference) ServiceOption {
 	return func(s *servingv1alpha1.Service) {
-		svcOwnerRefs := &s.OwnerReferences
-
-		switch {
-		case *svcOwnerRefs == nil:
-			*svcOwnerRefs = []metav1.OwnerReference{*or}
-		case metav1.GetControllerOf(s) == nil:
-			*svcOwnerRefs = append(*svcOwnerRefs, *or)
-		default:
-			for i, r := range *svcOwnerRefs {
-				if r.Controller != nil && *r.Controller {
-					(*svcOwnerRefs)[i] = *or
-				}
-			}
-		}
+		s.OwnerReferences = []metav1.OwnerReference{*or}
 	}
 }
 
-// WithContainerImage sets the container image of a Service. A minimal
-// Container definition is injected if the Service does not contain any.
+// WithContainerImage sets the container image of a Service.
 func WithContainerImage(img string) ServiceOption {
 	return func(s *servingv1alpha1.Service) {
-		if s.Spec.ConfigurationSpec.Template == nil {
-			s.Spec.ConfigurationSpec.Template = &servingv1alpha1.RevisionTemplateSpec{}
-		}
+		firstServiceContainer(s).Image = img
+	}
+}
 
-		containers := &s.Spec.ConfigurationSpec.Template.Spec.Containers
-		if *containers == nil {
-			*containers = make([]corev1.Container, 1)
-		}
-		(*containers)[0].Image = img
+// WithContainerPort sets the container port of a Service.
+func WithContainerPort(port int32) ServiceOption {
+	return func(s *servingv1alpha1.Service) {
+		ports := &firstServiceContainer(s).Ports
+
+		*ports = append(*ports, corev1.ContainerPort{
+			// empty name defaults to http/1.1 protocol
+			ContainerPort: port,
```

Comment:
please add a name for the port. Unnamed port cause weird issues with istio

Reply:
It's illegal in a Kn Service, unless you name it exactly "h2c" to force http2.

Reply:
oki doki

Reply:
No worries, Kn _does_ set a name on the Pod itself. This constructor only sets the port on the Kn Service, which gets converted to a deployment with all the missing stuff injected.